adam/learner/generics.py

```diff
@@ -0,0 +1,126 @@
+    ) -> LanguagePerceptionSemanticAlignment:
+        return language_perception_semantic_alignment
+
+    def enrich_during_description(
+        self, perception_semantic_alignment: PerceptionSemanticAlignment
+    ) -> PerceptionSemanticAlignment:
+        return perception_semantic_alignment
+
+    def log_hypotheses(self, log_output_path: Path) -> None:
+        logging.info(
+            "Logging %s hypotheses to %s",
+            len(self.learned_representations),
+            log_output_path,
+        )
+        Path(log_output_path).mkdir(parents=True, exist_ok=True)
+        with open(log_output_path / f"generics_log.txt", "w") as out:
+            for (
+                sequence,
+                (object_concept, others),
+            ) in self.learned_representations.items():
+                out.write(f'Learned generic: {" ".join(sequence)} \n')
+                out.write(f"Related objects: {object_concept} \n")
+                out.write(f"Related concept: {others} \n\n")
+
+    def templates_for_concept(self, concept: Concept) -> AbstractSet[SurfaceTemplate]:
+        return set()
+
+    def learn_from(
+        self,
+        language_perception_semantic_alignment: LanguagePerceptionSemanticAlignment,
+        offset: int = 0,
+    ) -> None:
+        sequence = (
+            language_perception_semantic_alignment.language_concept_alignment.language.as_token_sequence()
+        )
+        recognized_semantic_nodes = list(
+            language_perception_semantic_alignment.perception_semantic_alignment.semantic_nodes
+        )
+        span = (
+            language_perception_semantic_alignment.language_concept_alignment.node_to_language_span
+        )
+
+        # Get actions and attributes that are recognized in the scene
+        action_nodes = [
+            n for n in recognized_semantic_nodes if isinstance(n, ActionSemanticNode)
+        ]
+        attribute_nodes = [
+            n for n in recognized_semantic_nodes if isinstance(n, AttributeSemanticNode)
+        ]
+
+        # Check if a recognized object matches the heard utterance
+        significant_object_concept: Optional[Concept] = None
+        for node in recognized_semantic_nodes:
+            if isinstance(node, ObjectSemanticNode) and node in span:
+                significant_object_concept = node.concept
+
+        # Actions: E.g dog s walk
+        # Attributes: E.g cookies are brown
+        # Kinds: E.g cookie is a food
+        #  Both of these statements are regular generics as the noun is indefinite. We just need to
+        #  check attributes in addition to actions. We hardwire "are" in order to recognize kinds.
+        # If there is a recognized object node that matches the scene, and a generic action OR attribute, learn!
+        for nodes in [action_nodes, attribute_nodes]:
+            if significant_object_concept and nodes:
+                # Get the slot of object for each recognized action in the statement
```

Comment:
Wouldn't this be for each recognized action in the scene? (So if there were actions in the scene not mentioned in the statement, those would also be handled here.)

Reply:
That's right, and in fact that's been helpful because we now also generalize "sit down on" when we hear "sits". We don't typically hypothesize verbs incorrectly, so it's not an issue unless we introduce noise to generics, which is not our plan.

Reply:
Okay. Then I suggest changing the comment:
```suggestion
                # Get the slot of object for each recognized action in the scene
```
(I understood "statement" to mean "the linguistic description.")